kore/src/Kore/Step/BaseStep.hs

```diff
@@ -890,3 +863,526 @@ unwrapPredicateVariables
+        scatter unifiers
+
+{- | Instantiate the rule by applying the unification solution.
+
+The unification solution is normalized with the 'requires' clause from the
+unified rule. @instantiateRule@ fails if normalization fails. @instantiateRule@
+branches when the 'PredicateSubstitutionSimplifier' causes normalization to
+branch.
+
+ -}
+instantiateRule
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , FreshVariable  variable
+        , SortedVariable variable
+        )
+    => MetadataTools Object StepperAttributes
+    -> PredicateSubstitutionSimplifier Object
+    -> StepPatternSimplifier Object
+    -> BuiltinAndAxiomSimplifierMap Object
+
+    -> UnifiedRule variable
+    -- ^ Rule and unification solution
+    -> BranchT
+        (ExceptT (StepError Object variable) Simplifier)
+        (UnifiedRule variable)
+instantiateRule
+    metadataTools
+    predicateSimplifier
+    patternSimplifier
+    axiomSimplifiers
+
+    unified@Predicated { term = axiom }
+  = do
+    let unifier = unified { term = () }
+        RulePattern { requires, ensures } = axiom
+        merged =
+            PredicateSubstitution.fromPredicate requires
+            *> PredicateSubstitution.fromPredicate ensures
+            *> unifier
+    normalized <- normalize merged
+    let Predicated { substitution } = normalized
+        substitution' = Substitution.toMap substitution
+        axiom' = RulePattern.substitute substitution' axiom
+    return (normalized { term = axiom' })
+  where
+    normalize =
+        liftFromUnification
+        . Substitution.normalizeExcept
+            metadataTools
+            predicateSimplifier
+            patternSimplifier
+            axiomSimplifiers
+
+{- | Apply a rule to produce final configurations given some initial conditions.
+
+The rule should be instantiated with 'instantiateRule'. The initial conditions
+are merged with any conditions from the rule instantiation and
+normalized. @applyRule@ fails if normalization fails. @applyRule@ branches when
+the 'PredicateSubstitutionSimplifier' causes normalization to branch.
+
+ -}
+applyRule
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , FreshVariable  variable
+        , SortedVariable variable
+        )
+    => MetadataTools Object StepperAttributes
+    -> PredicateSubstitutionSimplifier Object
+    -> StepPatternSimplifier Object
+    -> BuiltinAndAxiomSimplifierMap Object
+
+    -> PredicateSubstitution Object variable
+    -- ^ Initial conditions
+    -> UnifiedRule variable
+    -- ^ Instantiated rule
+    -> BranchT
+        (ExceptT (StepError Object variable) Simplifier)
+        (ExpandedPattern Object variable)
+applyRule
+    metadataTools
+    predicateSimplifier
+    patternSimplifier
+    axiomSimplifiers
+
+    initial
+    instantiated
+  =
+    applyInitialConditions' initial (right <$> instantiated)
```

Comment:
Doesn't `instantiated` already contain the initial predicate+substitution?

Reply:
No, but I will refactor this section of code to make that clearer.